sources/packages/backend/apps/queue-consumers/src/services/cas-invoice-batch/cas-invoice-batch.service.ts

```diff
@@ -0,0 +1,254 @@
+            );
+            const student =
+              receipt.disbursementSchedule.studentAssessment.application
+                .student;
+            const offeringIntensity =
+              receipt.disbursementSchedule.studentAssessment.offering
+                .offeringIntensity;
+            // Create a new invoice for the receipt.
+            const newInvoice = new CASInvoice();
+            newInvoice.casInvoiceBatch = invoiceBatch;
+            newInvoice.disbursementReceipt = receipt;
+            newInvoice.casSupplier = student.casSupplier;
+            newInvoice.invoiceNumber = `SIMS-INVOICE-${newInvoiceSequence++}-${
+              student.casSupplier.supplierNumber
+            }`;
+            newInvoice.invoiceStatus = CASInvoiceStatus.Pending;
+            newInvoice.invoiceStatusUpdatedOn = now;
+            newInvoice.creator = this.systemUsersService.systemUser;
+            // Create invoice details.
+            newInvoice.casInvoiceDetails = this.createInvoiceDetails(
+              receipt.disbursementSchedule.disbursementValues,
+              offeringIntensity,
+              distributionAccounts,
+            );
+            casInvoices.push(newInvoice);
+            invoicesSummary.info(
+              `Invoice ${newInvoice.invoiceNumber} created for receipt ID ${receipt.id}.`,
+            );
+            newInvoice.casInvoiceDetails.forEach((detail) => {
+              invoicesSummary.info(
+                `Created invoice detail for award ${detail.casDistributionAccount.awardValueCode}(${detail.casDistributionAccount.operationCode}).`,
+              );
+            });
+          }
+          return newInvoiceSequence;
+        },
+      );
+      await entityManager.getRepository(CASInvoice).save(casInvoices, {
+        chunk: INVOICES_CHUNK_SIZE_INSERTS,
+      });
+      // Associate the created invoices to the batch to be returned.
+      invoiceBatch.casInvoices = casInvoices;
+      return invoiceBatch;
+    });
+  }
+
+  /**
+   * Creates invoice details to be associated with an invoice to be saved to the database.
+   * @param disbursedAwards disbursed awards to be used for creating the invoice details.
+   * @param offeringIntensity offering intensity used to filter the distribution accounts.
+   * @param distributionAccounts list of all active distribution accounts to allow the creation of the
+   * invoice details for each disbursed award.
+   * @returns invoice details to be associated with an invoice.
+   */
+  private createInvoiceDetails(
+    disbursedAwards: Pick<DisbursementValue, "valueCode" | "effectiveAmount">[],
+    offeringIntensity: string,
+    distributionAccounts: CASDistributionAccount[],
+  ): CASInvoiceDetail[] {
+    // New invoice creation.
+    const invoiceDetails: CASInvoiceDetail[] = [];
+    for (const disbursedAward of disbursedAwards) {
+      // Find the distribution accounts for the award.
+      const accounts = distributionAccounts.filter(
```

Comment:
Will it not be useful to throw error if the distribution account size is 0?

Reply:
It may, but I will raise the question and leave it to a second PR.
It may be useful if it is 0 records, or if there is only one, or if it is not a pair of debit and credit operations or they may want the ability to not have any and avoid including it in the invoice.

Reply:
It will be taken care on an upcoming PR as per business agreement. Just some basic validation.